### PR TITLE
Release 1.4

### PR DESCRIPTION
**Pull request by martijnbastiaan**
I'd like to include:

* [x] https://github.com/clash-lang/clash-compiler/pull/1680
* [x] https://github.com/clash-lang/clash-compiler/pull/1676 
* [x] https://github.com/clash-lang/clash-compiler/pull/1682 (fix for an issue introduced by https://github.com/clash-lang/clash-compiler/commit/e16e9012a19b0296bf0ace7c23ce44c200060e03)

But neither is crucial. If they don't make it in, we'll just include them in 1.6.

### CODE REVIEW

File: clash-prelude/src/Clash/Class/AutoReg/Internal.hs [hunk 1]
```diff
@@ -48,7 +48,12 @@ import           Language.Haskell.TH.Syntax
 import           Language.Haskell.TH.Lib
 import           Language.Haskell.TH.Ppr
 
-import           Control.Lens.Internal.TH     (bndrName, conAppsT)
+import           Control.Lens.Internal.TH     (conAppsT)
+
+-- | Return 'Name' contained in a 'TyVarBndr'.
+bndrName :: TyVarBndr -> Name
+bndrName (PlainTV  n  ) = n
+bndrName (KindedTV n _) = n
```
Comment (christiaanb) [inline clash-prelude/src/Clash/Class/AutoReg/Internal.hs:56]:
```suggestion
#if MIN_VERSION_lens(5,1,0)
import           Control.Lens.Internal.TH     (bndrName, conAppsT)
#else
import           Control.Lens.Internal.TH     (conAppsT)
#endif

#if !MIN_VERSION_lens(5,1,0)
-- | Return 'Name' contained in a 'TyVarBndr'.
bndrName :: TyVarBndr -> Name
bndrName (PlainTV  n  ) = n
bndrName (KindedTV n _) = n
#endif
```

**Review by christiaanb — CHANGES_REQUESTED** [review]

File: clash-prelude/src/Clash/Class/AutoReg/Internal.hs (hunk 1 above)
Comment (christiaanb) [reply clash-prelude/src/Clash/Class/AutoReg/Internal.hs:56]: Misunderstood the intention

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,74 @@
...
+  * Type arguments of `Clash.Sized.Vector.fold` swapped: before `forall a n . (a -> a -> a) -> Vec (n+1) a -> a`, after `forall n a . (a -> a -> a) -> Vec (n+1) a`. This makes it easier to use `fold` in a `1 <= n` context so you can "simply" do `fold @(n-1)`
+  * Moved `Clash.Core.Evaluator` into `Clash.GHC` and provided generic interface in `Clash.Core.Evalautor.Types`. This removes all GHC specific code from the evaluator in clash-lib.
+  * Clash no longer disables the monomorphism restriction. See [#1270](https://github.com/clash-lang/clash-compiler/issues/1270), and mentioned issues, as to why. This can cause, among other things, certain eta-reduced descriptions of sequential circuits to no longer type-check. See [#1349](https://github.com/clash-lang/clash-compiler/pull/1349) for code hints on what kind of changes to make to your own code in case it no longer type-checks due to this change.
```
Comment (christiaanb) [inline CHANGELOG.md:31]: I think we should highlight this specifically. This is going to break exisiting code.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,4 +1,74 @@
...
+  * Signals on different domains used to be coercable because the domain had a type role "phantom". This has been changed to "nominal" to prevent accidental, unsafe coercions. [#1640](https://github.com/clash-lang/clash-compiler/pull/1640)
+  * Size parameters on types in Clash.Sized.Internal.* are now nominal to prevent unsafe coercions. [#1640](https://github.com/clash-lang/clash-compiler/pull/1640)
+  * `hzToPeriod` now takes a `Ratio Natural` rather than a `Double`. It rounds slightly differently, leading to more intuitive results and satisfying the requested change in [#1253](https://github.com/clash-lang/clash-compiler/issues/1253). Clash expresses clock rate as the clock period in picoseconds. If picosecond precision is required for your design, please use the exact method of specifying a clock period rather than a clock frequency.
+  * `periodToHz` now results in a `Ratio Natural`
+
+
+Added:
+  * Support for GHC 9.0.1
+  * 'Clash.Signal.sameDomain': Allows user obtain evidence whether two domains are equal.
```
Comment (DigitalBrains1) [inline CHANGELOG.md:43]: These are the wrong quotes :-). This is Markdown, not Haddock :-).

Comment (alex-mckenna) [reply CHANGELOG.md:43]: :fish: :x:

File: CHANGELOG.md (hunk 2 above)
Comment (martijnbastiaan) [reply CHANGELOG.md:31]: I'll make an additional category "highlighted changes".

File: CHANGELOG.md (hunk 3 above)
Comment (martijnbastiaan) [reply CHANGELOG.md:43]: I changed sooo many already :')

**Pull request merged**